src/Umbraco.Web/PublishedCache/NuCache/DataSource/IDataSource.cs

```diff
@@ -73,5 +75,77 @@ internal interface IDataSource
         /// MUST be ordered by level + parentId + sortOrder!
         /// </remarks>
         IEnumerable<ContentNodeKit> GetTypeMediaSources(IScope scope, IEnumerable<int> ids);
+
+        /// <summary>
+        /// Removes all versions of an entity from the cache (both published and current
+        /// </summary>
+        /// <param name="id"></param>
+        void RemoveEntity(IScope scope, int id);
```

Comment:
These changes are going to be a complete nightmare to migrate to netcore since this has already been split apart and done in a completely different manner. To make this feasible I think this PR is going to need to be split into different ones:

* Fix binary bulk insert for SQLCE and ordering of properties.
* Fix JsonSerializerSettings instantiation
* Fix issue where Guid was not being mapped to Key in SQL Select for content types, it was being mapped to UniqueId.
* Add component to automatically rebuild the cmsContentNu table when switching serializers.
* Use ITransactableDictionary to decouple from bplustree

Those should be relatively easy, the last one is harder

* Move SQL from PublishedSnapshotService into DatabaseDataSource

For the DatabaseDataSource changes, I'm just not sure the best way forward with merging to netcore. Might be easier to just leave it as-is knowing that in netcore it's changed?

Reply:
I'm happy to split this up. Without making things public and moving the sql into a repository, it's going to be difficult to swap out sql for another storage provider.

Reply:
Fix issue where Guid was not being mapped to Key in SQL Select for content types, it was being mapped to UniqueId.
https://github.com/umbraco/Umbraco-CMS/pull/9804

Reply:
Fix binary bulk insert for SQLCE and ordering of properties.
https://github.com/umbraco/Umbraco-CMS/pull/9805

Reply:
Fix JsonSerializerSettings instantiation
https://github.com/umbraco/Umbraco-CMS/pull/9670

Reply:
auto rebuild nucache
https://github.com/umbraco/Umbraco-CMS/pull/9806

Reply:
> For the DatabaseDataSource changes, I'm just not sure the best way forward with merging to netcore. Might be easier to just leave it as-is knowing that in netcore it's changed?

The refactor is still necessary for more performance improvements in v8, so my recommendation would be to refactor in v8 and leave .net core as is or if possible, backport the changes from .net core

Reply:
Use ITransactableDictionary to decouple from bplustree
Move SQL from PublishedSnapshotService into DatabaseDataSource
https://github.com/umbraco/Umbraco-CMS/pull/9857